analyzers/src/SonarAnalyzer.CSharp/Rules/PointersShouldBePrivate.cs

```diff
@@ -53,25 +53,38 @@ protected override void Initialize(SonarAnalysisContext context)
 
                 foreach (var variable in fieldDeclaration.Declaration.Variables)
                 {
-                    var variableSymbol = (IFieldSymbol)c.SemanticModel.GetDeclaredSymbol(variable);
-
-                    if (variableSymbol != null &&
-                        variableSymbol.Type.IsAny(KnownType.PointerTypes))
+                    if (SymbolIfPointerType(fieldDeclaration.Declaration, variable, c.SemanticModel) is { } variableSymbol
+                        && variableSymbol.GetEffectiveAccessibility() is var accessibility
+                        && accessibility != Accessibility.Private
+                        && accessibility != Accessibility.Internal
+                        && !variableSymbol.IsReadOnly)
                     {
-                        var accessibility = variableSymbol.GetEffectiveAccessibility();
-                        if (accessibility == Accessibility.Private ||
-                            accessibility == Accessibility.Internal ||
-                            variableSymbol.IsReadOnly)
-                        {
-                            return;
-                        }
-
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, variable.GetLocation(),
-                            variableSymbol.Name));
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, variable.GetLocation(), variableSymbol.Name));
                     }
                 }
             },
             SyntaxKind.FieldDeclaration);
+
+        private static IFieldSymbol SymbolIfPointerType(VariableDeclarationSyntax variableDeclaration, VariableDeclaratorSyntax variableDeclarator, SemanticModel semanticModel)
+        {
+            if (variableDeclaration.Type.IsKind(SyntaxKind.PointerType)
+                || IsUnmanagedFunctionPointer(variableDeclaration))
+            {
+                return (IFieldSymbol)semanticModel.GetDeclaredSymbol(variableDeclarator);
+            }
+            else
+            {
+                return ((IFieldSymbol)semanticModel.GetDeclaredSymbol(variableDeclarator)) is { } variableSymbol
```

Comment:
you can do an AST-based check on `variableDeclaration.Type` to be / contain `IntPtr` or `UIntPtr`, to avoid calling the semantic model for each field symbol?